uaclient/tests/test_lock.py

```diff
@@ -0,0 +1,219 @@
+class TestSingleAttemptLock:
+    def test_creates_and_releases_lock(
+        self,
+        m_write_cache,
+        m_add_notice,
+        m_delete_cache_key,
+        _m_getpid,
+        FakeConfig,
+    ):
+        cfg = FakeConfig()
+        arg, kwarg = mock.sentinel.arg, mock.sentinel.kwarg
+
+        def test_function(args, cfg):
+            assert arg == mock.sentinel.arg
+            assert kwarg == mock.sentinel.kwarg
+            return mock.sentinel.success
+
+        with SingleAttemptLock(cfg, "some operation"):
+            ret = test_function(arg, cfg=cfg)
+
+        assert mock.sentinel.success == ret
+        assert [
+            mock.call("lock", "123:some operation")
+        ] == m_write_cache.call_args_list
+        lock_msg = "Operation in progress: some operation"
+        assert [mock.call("", lock_msg)] == m_add_notice.call_args_list
+        assert [mock.call("lock")] == m_delete_cache_key.call_args_list
+
+    def test_creates__and_releases_lock_when_error_occurs(
+        self,
+        m_write_cache,
+        m_add_notice,
+        m_delete_cache_key,
+        _m_getpid,
+        FakeConfig,
+    ):
+        cfg = FakeConfig()
+
+        def test_function():
+            raise RuntimeError("test")
+
+        with pytest.raises(RuntimeError) as exc:
+            with SingleAttemptLock(cfg, "some operation"):
+                test_function()
+
+        assert "test" == str(exc.value)
+        assert [
+            mock.call("lock", "123:some operation")
+        ] == m_write_cache.call_args_list
+        lock_msg = "Operation in progress: some operation"
+        assert [mock.call("", lock_msg)] == m_add_notice.call_args_list
+        assert [mock.call("lock")] == m_delete_cache_key.call_args_list
+
+    @mock.patch(M_PATH_UACONFIG + "check_lock_info", return_value=(10, "held"))
+    def test_raises_lock_held_when_held(
+        self,
+        m_write_cache,
+        m_add_notice,
+        m_delete_cache_key,
+        _m_getpid,
+        _m_check_lock_info,
+        FakeConfig,
+    ):
+        cfg = FakeConfig()
+        arg, kwarg = mock.sentinel.arg, mock.sentinel.kwarg
+
+        def test_function(args, cfg):
+            assert arg == mock.sentinel.arg
+            assert kwarg == mock.sentinel.kwarg
+            return mock.sentinel.success
+
+        with pytest.raises(LockHeldError) as exc:
+            with SingleAttemptLock(cfg, "some operation"):
+                test_function(arg, cfg=cfg)
+
+        assert exc.value.lock_holder == "held"
+        assert exc.value.lock_holder_pid == 10
+        assert exc.value.lock_request == "some operation"
+
+        assert [
+            mock.call("lock", "123:some operation")
+        ] != m_write_cache.call_args_list
+        lock_msg = "Operation in progress: some operation"
+        assert [mock.call("", lock_msg)] != m_add_notice.call_args_list
+        assert [mock.call("lock")] != m_delete_cache_key.call_args_list
+
+
+class TestSpinLock:
```

Comment:
The first two tests of `TestSpinLock`  are pretty similar to the ones one `TestSingleAttemptLock`
What do you think we have a common lock test class that we can parametrize which lock class we are using ?

We can that just use the `TestSpinLock` for the specific functionality of that class

Reply:
Good idea, I'll try it out. I don't want to complicate the tests too much just to remove a bit of duplication though

Reply:
I had a similar idea but want to run it by you before implementing @lucasmoura.

Would it make sense for the `SpinLock` class tests to only verify that it calls `SingleAttemptLock.__enter__/__exit__` instead of testing that it calls the correct `write_cache/add_notice/delete_cache` functions? That would simplify the SpinLock tests a bit

Reply:
But if we do that, then we would lose a little bit of our tracking of what is happening inside both of these methods right ?

Reply:
Right, actually I think it would just look like deleting `test_creates_lock_and_notice` and `test_creates_lock_and_notice_when_error_occurs`. On second thought I don't think we should do that. I'll do the refactor you suggest

Reply:
@lucasmoura updated! Sorry for the FUD, that was straightforward, and is definitely better